airbyte-integrations/connectors/source-snapchat-marketing/source_snapchat_marketing/source.py

```diff
@@ -0,0 +1,354 @@
+
+
+class Creatives(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-creatives """
+
+    data_field = ["creatives", "creative"]
+    cls_object = AdAccounts
+
+
+class Media(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-media """
+
+    data_field = ["media", "media"]
+    cls_object = AdAccounts
+
+
+class Campaigns(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-campaigns """
+
+    data_field = ["campaigns", "campaign"]
+    cls_object = AdAccounts
+
+
+class Ads(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-ads-under-an-ad-account """
+
+    data_field = ["ads", "ad"]
+    cls_object = AdAccounts
+
+
+class Adsquads(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-ad-squads-under-an-ad-account """
+
+    data_field = ["adsquads", "adsquad"]
+    cls_object = AdAccounts
+
+
+class Segments(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-audience-segments """
+
+    data_field = ["segments", "segment"]
+    cls_object = AdAccounts
+
+
+class SnapchatAdsOauth2Authenticator(Oauth2Authenticator):
+    """Request example for API token extraction:
+    curl -X POST https://accounts.snapchat.com/login/oauth2/access_token \
+      -d "refresh_token={refresh_token}" \
+      -d "client_id={client_id}" \
+      -d "client_secret={client_secret}"  \
+      -d "grant_type=refresh_token"  \
+    """
+
+    def __init__(self, config):
+        super().__init__(
+            token_refresh_endpoint="https://accounts.snapchat.com/login/oauth2/access_token",
+            client_id=config["client_id"],
+            client_secret=config["secret"],
+            refresh_token=config["refresh_token"],
+        )
+
+    def refresh_access_token(self) -> Tuple[str, int]:
+        response_json = None
+        try:
+            response = requests.request(method="POST", url=self.token_refresh_endpoint, data=self.get_refresh_request_body())
+            response_json = response.json()
+            response.raise_for_status()
+            return response_json["access_token"], response_json["expires_in"]
+        except requests.exceptions.RequestException as e:
+            if response_json and "error" in response_json:
+                raise Exception(
+                    "Error refreshing access token. Error: {}; Error details: {}; Exception: {}".format(
+                        response_json["error"], response_json["error_description"], e
+                    )
+                ) from e
+            raise Exception(f"Error refreshing access token: {e}") from e
+
+
+# Source
+class SourceSnapchatMarketing(AbstractSource):
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
```

Comment:
Could you add a docstring comment to a source with what Snapchat Marketing does. It helps understand what happened in code on a high level.  

For example: `"""Source Intercom fetch data from messaging platform."""`

Reply:
Done